### ISSUE TITLE
Move Sse2.SetZeroVector128<T> to Sse class

### ISSUE THREAD

**Issue by fiigii** [dotnet/coreclr#16575]
We should move `Sse2.SetZeroVector128<T>` to `Sse` class to keep one generic intrinsic and match AVX API design.

cc @CarolEidt @tannergooding @eerhardt @mikedn @4creators

**Comment by tannergooding** [comment]
I actually think we should keep them split and explode them instead.

`Sse` only deals with floats and you don't get any way to actually "consume" the other types until `Sse2`.

If you really need to zero a vector of another type when using `Sse`, you can explicitly call `SetZeroVector128Single` and then `StaticCast`.

I likewise think that `AVX` should be split, since `AVX` only deals with `float` and `double` and `AVX2` deals with the 8 integer types.

**Comment by tannergooding** [comment]
I believe this also ensures the user is aware of the "data bypass penalty" mentioned here: https://github.com/dotnet/coreclr/pull/16183#discussion_r170430816

Where-as it may not be obvious otherwise.

**Comment by fiigii** [comment]
I think we can provide the semantics of `SetZeroVector128<T>` in `Sse` and hepler-intrinsic codgen usually is not "fixed". That would simplify the API and JIT compiler.

> I believe this also ensures the user is aware of the "data bypass penalty" 

Actually, the "data bypass penalty" is not very significant that is just 1-cycle latency on most of x86 architectures.

**Comment by tannergooding** [comment]
> I think we can provide the semantics of SetZeroVector128<T> in Sse and hepler-intrinsic codgen usually is not "fixed"

Perhaps it is worth a more "in-depth" discussion where these helpers live. They are not tied, technically speaking, to any given ISA but due to where they are exposed are being "tied down". They additionally are enforcing a "generational" dependency which isn't documented (but is instead just widely assumed -- I wasn't able to find any explicit documentation in the architecture's manual).

**Comment by 4creators** [comment]
> We should move Sse2.SetZeroVector128<T> to Sse class to keep one generic intrinsic and match AVX API design.

The problem is that this proposal is against the very basic principle of the design where we keep intrinsics bound to the ISA class where underlying instructions are available and where C++ intrinsics were put in native compilers even for helper intrinsics. Every developer having previous experience with SIMD programming would expect to find them splitted between Sse and Sse2 class.

**Comment by fiigii** [comment]
> Every developer having previous experience with SIMD programming would expect to find them splitted between Sse and Sse2 class.

This is the current design goal. That splits `SetZeroVector128` in Sse and Sse2, but define `SetZeroVector256<T>` in Avx, which matches C++ intrinsic design.

**Comment by tannergooding** [comment]
We have already deviated from C++ on a number of things, such as naming, the shape of the APIs, etc. Using the C/C++ intrinsics as a general basis/starting point is definitely good, but we should continue to ensure that what we expose and where we expose it makes sense for .NET

This is a slightly larger discussion and definitely needs input from people like @eerhardt, who can better guide the shape this API should ultimately have.

In my opinion, we should:
* Ensure these are exposed someplace logical
*  Will the user be able to guess where it is?
* Try to minimize, or at the very least document, any "gotchas"
  * `StaticCast` is exposed in `Sse`, but you can't really use the result unless `Sse2` is also supported, since the `Sse` APIs only consume `float`
  * `SetZeroVector256<T>` is exposed in `Avx`, but you can't use integer results unless `Avx2` is also supported and it my incur a "data bypass penalty"
  * etc
* Ensure we are consistent

**Comment by 4creators** [comment]
> This is a slightly larger discussion and definitely needs input from people like @eerhardt 

I fully agree. IMO it would be helpful to get input from devs who are less biased than we are after couple of months of HW intrinsics implementation work

**Comment by mikedn** [comment]
> StaticCast is exposed in Sse, but you can't really use the result unless Sse2 is also supported, since the Sse APIs only consume float

Since this one doesn't actually do anything maybe it shouldn't be connected in anyway to the intrinsic classes. Maybe it should be a member of the vector structs. Something like `Vector128<T> AsVector128<T>()`. Though I suspect it may be problematic for the JIT to recognize that.

**Comment by tannergooding** [comment]
> It's the natural way to do things in C#.

Only until https://github.com/dotnet/csharplang/issues/99 is approved (if that ever happens) 😄 

> I'd seriously consider removing the "Set zero" intrinsics and rely on struct constructors to generate zero vectors

I believe we made an explicit design decision to keep these opaque.

However, it does make sense to me that `new Vector128<T>()` would "play nicely" and ultimately produce the "optimal" codegen. There would still be some open questions about how it would behave for Vector128<T> when `T` is not supported by the underlying platform, however.

That being said, I also don't think that we should remove the set zero intrinsics themselves, it helps with discoverability and porting of existing code. It also is easier to provide "concrete" guarantees about the given intrinsic.

**Comment by mikedn** [comment]
> There would still be some open questions about how it would behave for Vector128 when T is not supported by the underlying platform, however.

Funny:
```C#
struct Num
{
    public int a;
}

[MethodImpl(MethodImplOptions.NoInlining)]
static Vector128<int> Test()
{
    Vector128<Num> v1 = new Vector128<Num>();
    Vector128<Num> v2 = new Vector128<Num>();
    return Sse2.Add(Sse.StaticCast<Num, int>(v1), Sse.StaticCast<Num, int>(v2));
}

```
asserts:
```
Assert failure(PID 12524 [0x000030ec], Thread: 8692 [0x21f4]): Assertion failed 'srcType != TYP_UNKNOWN' in 'Program:Test():struct' (IL size 34)

    File: d:\projects\coreclr\src\jit\hwintrinsicxarch.cpp Line: 478
    Image: D:\Projects\coreclr\bin\Product\Windows_NT.x64.Checked\CoreRun.exe
```

**Comment by fiigii** [comment]
@mikedn Thanks, will fix.

**Comment by eerhardt** [comment]
Can we come up with a holistic proposal to address these methods?  Taking a cursory look at the methods, there are some inconsistencies between:

```
Sse - public static Vector128<float> SetZeroVector128() 
Sse - public static Vector128<float> SetAllVector128(float value)

Sse2 - public static Vector128<T> SetZeroVector128<T>()
Sse2 - public static Vector128<byte> SetAllVector128(byte value) [[Exploded]]

Avx - public static Vector256<T> SetZeroVector256<T>()
Avx - public static Vector256<T> SetAllVector256<T>(T value)
```

1. I thought [the plan](https://github.com/dotnet/designs/pull/28/files#diff-980684dba5fee744c5d7c697229a4d4bR105) was to only use a generic method when the generic method supported all the numeric types.  `Avx` only supports `float` and `double`, so it seems incorrect to have generic methods here.
2. Why does `Sse2.SetZeroVector128` use a generic type, but `Sse2.SetAllVector128` explode out for each type?  Similarly, `Avx.SetAllVector256` uses a generic type, but `Sse2` doesn't .

Then, looking further at `Avx.LoadAlignedVector256`, this supports all 10 numeric types, but has the types exploded.  Why should this method be exploded, but not `SetZeroVector256` and `SetAllVector256`?

**Comment by tannergooding** [comment]
My preference, for Sse/Sse2, would be to:
* Keep `Sse2.SetZeroVector128` in `Sse2`, but explode it
* Rename `SetZeroVector128` to match the current naming convention (i.e. `Sse.SetZeroVector128` should be `Sse.SetZeroVector128Single`, and `Sse2.SetZeroVector128` should be `Sse2.SetZeroVector128Double`, etc)

For AVX, I would like to better understand why we have certain intrinsics exposed there.
* AVX, in general, only supports `float` and `double`.
* The exception to this is `Load/Store` where AVX also exposes `vmovdqu` and `vmovdqa`.
* However, we are also exposing `Extract`/`Insert`/`SetAllVector`/`SetZeroVector` intrinsics for the integer types.

For most of these additional intrinsics, they are listed as "helper" and the code they generate is not clearly defined. I think it would be good to indicate what code we plan on having them actually use and if there is any penalty for that code (even if that "penalty" is only a 1-cycle latency).

For several of the `Extract`/`Insert` overloads, they are actually listed as using `vextractf128`/`vinsertf128` which are specified as operating over packed floating-point data.

It would be good to understand why these should be exposed here, rather than AVX2:
* What scenarios this will actually support
* Is this code actually typical
* Is it worthwhile making the performance penalty, if one exists,  more explicit
* etc
